File: backend/internal/domain/maintenance.go
package domain

import (
	"context"
	"time"
)

type MaintenanceWindow struct {
	ID          string     `json:"id"`
	ServiceID   string     `json:"service_id"`
	Title       string     `json:"title"`
	Description *string    `json:"description"`
	StartsAt    time.Time  `json:"starts_at"`
	EndsAt      time.Time  `json:"ends_at"`
	CreatedAt   time.Time  `json:"created_at"`
	ArchivedAt  *time.Time `json:"archived_at"`
}

type CreateMaintenanceWindowParams struct {
	ServiceID   string
	Title       string
	Description *string
	StartsAt    time.Time
	EndsAt      time.Time
}

type UpdateMaintenanceWindowParams struct {
	Title       *string
	Description *string
	StartsAt    *time.Time
	EndsAt      *time.Time
}

type MaintenanceWindowRepository interface {
	Create(ctx context.Context, p CreateMaintenanceWindowParams) (*MaintenanceWindow, error)
	GetByID(ctx context.Context, id string) (*MaintenanceWindow, error)
	ListByService(ctx context.Context, serviceID string) ([]*MaintenanceWindow, error)
	ListActive(ctx context.Context) ([]*MaintenanceWindow, error)
	ListActiveForService(ctx context.Context, serviceID string) ([]*MaintenanceWindow, error)
	IsActiveForService(ctx context.Context, serviceID string) (bool, error)
	Update(ctx context.Context, id string, p UpdateMaintenanceWindowParams) (*MaintenanceWindow, error)
	Archive(ctx context.Context, id string) error
}

File: frontend/lib/api/maintenance.ts
import { api } from "./client";
import type { MaintenanceWindow } from "@/types";

export const maintenanceApi = {
  list: (serviceId?: string) =>
    api.get<MaintenanceWindow[]>(`/admin/maintenance-windows${serviceId ? `?service_id=${serviceId}` : ""}`),
  get: (id: string) => api.get<MaintenanceWindow>(`/admin/maintenance-windows/${id}`),
  create: (data: {
    service_id: string;
    title: string;
    description?: string;
    starts_at: string;
    ends_at: string;
  }) => api.post<MaintenanceWindow>("/admin/maintenance-windows", data),
  update: (
    id: string,
    data: { title?: string; description?: string; starts_at?: string; ends_at?: string }
  ) => api.patch<MaintenanceWindow>(`/admin/maintenance-windows/${id}`, data),
  archive: (id: string) => api.delete<void>(`/admin/maintenance-windows/${id}`),
};
